src/Microsoft.VisualStudio.Threading/AsyncLazy`1.cs

```diff
@@ -183,7 +207,12 @@ public Task<T> GetValueAsync(CancellationToken cancellationToken)
                         }
                     };
 
-                    this.recursiveFactoryCheck.Value = RecursiveCheckSentinel;
+                    if (!this.SuppressRecursiveFactoryDetection)
+                    {
+                        Assumes.Null(this.recursiveFactoryCheck);
+                        this.recursiveFactoryCheck = new AsyncLocal<object>() { Value = RecursiveCheckSentinel };
+                    }
+
```

Comment:
Is there a race here?

Reply:
I don't think so. There's a lock on L181, and only one thread ever reaches this point.
But if you think I'm missing something, do please share.

Reply:
I missed the lock hiding in the collapsed code region, sorry.  Nothing to see here, carry on.